ES6 Classes

heff opened the pull request:
<strike>This is built on #1976 so the commit to look at is https://github.com/videojs/video.js/commit/b7f6934d0cc4b6262f5c9ab6751e337a1259d300</strike>
An easier view of the changes here: https://github.com/mmcc/video.js/pull/7

So far I just converted Component to an ES6 class, and then re-added `init` and `extend` so that all the subclasses would still work (all tests are passing). After fully implementing this we would't have either of those methods anymore.

### Important notes
1. ES6 classes (through babelify) requires an Object.create shim. Before this we don't yet need to include the shim/sham combo, but after this we would.
2. This will remove the `MyComponent.extend()` static method, for all components, techs, etc., breaking that major API. We can't automatically add the static extend() method in the ES6 inheritance process like we're doing now for every sub class. We can manually add it for each component if we think that's important for backwards compatibility, maybe at least for techs. Otherwise we should introduce a `videojs.extends/videojs.inherits` for external projects that don't use ES6 classes. We can even just use Node's.

Please let me know any thoughts on this.

heff commented:
So apparently static methods are inherited in ES6 classes, so that makes the old `extend()` API really easy to add since it only needs to be added once.

Also I'm using the latest version of babelify but it's yelling at me for something I think it shouldn't.
```
createEl() {
  return super();
}

SyntaxError: slider/slider.js: Line 48: Direct super call is illegal in non-constructor, use super."createEl"() instead while parsing file
```

I instead have to do
```
createEl() {
  return super.createEl();
}
```

The first way works in the babel demo interface, so I'm not sure what's going on there.

gkatsev commented:
I think the spec changed to be the second example. I can find out for sure later.

heff commented:
The ES6 class work has been updated and all classes have been converted. All tests are passing so it could be merged in. I broke out any individual class into its own file, and some of the library got reorganized because of that.

This was mostly a syntax and organization change, with almost no functionality change. However, there's a limitation where you can't use `this` before `super` in a constructor, and I had to change when some variables are set. I was pretty careful about the order of things so I don't believe anything functionality should have changed around that.

However, I've realized we should be using `options` more to pass info to super classes, instead of setting values on the prototype. For example we have a bunch of `featuresX` properties on techs, and we really should be passing that through options.

I also renamed MediaTechController to Tech and the /media/ directory to /tech/ for clarity. But I also aliased Tech to MediaTechController so old techs won't break on that alone.

I have not converted over all static methods, and not sure if we need to.

Also, this and #2015...
![1315301583_sudden_headon_collision](https://cloud.githubusercontent.com/assets/166/7035535/95d9a8f4-dd57-11e4-9c4b-530b4b385484.gif)

mmcc commented:
![hide](http://media.giphy.com/media/12yBKshqZ28MRG/giphy.gif)

mmcc commented on src/js/big-play-button.js:
```diff
@@ -11,20 +10,21 @@ import Button from './button';
 
-export default BigPlayButton;
+Button.registerComponent('BigPlayButton', BigPlayButton);
+export default BigPlayButton;
```
Am I reading this right that this is removing the trailing newline? I won't comment on every instance of this, but I feel as strongly about a trailing newline as I do about removing empty white space.

mmcc replied:
Which is to say, very strongly. I'm pretty sure revolutions have been fought over the need for trailing newlines.

heff replied:
Ha, I guess that's what it is. And I guess Atom isn't configured to keep those for me. I'll update it.

mmcc replied:
Looks like it's [enabled by default now](https://github.com/atom/atom/issues/399), which would be weird if you're not seeing it.

mmcc commented on src/js/button.js:
```diff
@@ -27,64 +24,65 @@ var Button = Component.extend({
+      tabIndex: 0
+    }, props);
+
+    let el = Component.prototype.createEl.call(this, type, props);
```
Isn't `Component` `super` in this case?

heff replied:
Yeah, I didn't do the most exhaustive search of old super method use cases, except for in the constructor. I guess I should do that.

mmcc commented on src/js/button.js:
```diff
@@ -27,64 +24,65 @@ var Button = Component.extend({
-Button.prototype.onFocus = function(){
-  Events.on(document, 'keydown', Lib.bind(this, this.onKeyPress));
-};
+  onFocus(){
```
Nitpick: let's be consistent about spacing. `buildCSSClass() {` vs `onFocus(){`. I personally prefer the space, but let's just do it one way.

heff replied:
Yeah, I'm good with the space. I'd love to use escodegen to just rewrite the whole lib with the right formatting. Maybe after all of this is done.

gkatsev replied:
Yep, spacing. It would be interesting to have a pre-commit hook that modifying the code styles based on output from node-jscs and jshint/eslint.
Also, I'm still debating whether we should have a space before `()` in the concise methods.
i.e.
```js
{
  foo() {}
}
// vs
{
  foo () {}
}
```

For function statements, we're supposed to put a space after the name and this sets it apart from where we call the function and where we declare the function. The same could be applied to the concise methods.

mmcc commented on src/js/button.js:
```diff
@@ -27,64 +24,65 @@ var Button = Component.extend({
+}
 
+
+Component.registerComponent('Button', Button);
```
:+1: I like the idea of registering components and exports all at the bottom of each module.

heff replied:
Cool. One interesting thing that came out of this is every component has an inherited `registerComponent` function, so we don't have to include Component just to register something. Though I couldn't decide if I should register the component with it's own register function or its super classes.
```
Button.registerComponent('PlayToggle', PlayToggle);
or
PlayToggle('PlayToggle', PlayToggle);
```

heff replied:
Sorry, chopped off part of the example. It's fixed now.

```
Button.registerComponent('PlayToggle', PlayToggle);
or
PlayToggle.registerComponent('PlayToggle', PlayToggle);
```

mmcc commented:
This isn't a problem with the PR itself, obviously, but this diff is useless.

mmcc commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+    let parentOptions = parent.options();
+    let children = parentOptions['children'];
+
+    let handleAdd;
```
Why is this defined outside of the `if` block if it's only used within? The only other path is the else, which could easily be changed to `Lib.obj.each(children, null);` and honestly be more readable.

mmcc replied:
Wait a second, think the diff was screwing with me there, it doesn't look like there's *any* reason to define this outside of the `if` block.

heff replied:
It made sense when it was vars. The if block used children which was defined with parentOptions which was defined with parent. I didn't actually touch this part but I've cleaned it up for you. :)

mmcc commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   * @return {vjs.Component}        self
+   */
+  on(first, second, third){
+    var target, type, fn, removeOnDispose, cleanRemover, thisComponent;
```
Can we decide what we want to do about this from a style perspective? I, for one, **much** prefer to define a variable at the point of use or as close as possible to it, but I understand this is a bit of stylistic taste.

FWIW, I found *quite* a few unused variables when I did the first ES6 pass, and this style is inherently prone to that kind of thing.

gkatsev replied:
I've been coming around on using the variables where needed with let. Also, only having on variable per declaration.
However, if possible, declaring all variables at the top of the scope/block is still preferable.

gkatsev replied:
Also, it would be nice to move all this `on` stuff into the `EventEmitter` class we have and then have Component extend from it or something like that.

heff commented:
Yeah, the diff is bad. I could have avoided moving files for the first pass, but even in the files I didn't move, the complete restructuring of the class makes it impossible to tell what exactly changed.

mmcc commented:
It's a lot better if you switch to the split diff, so that's what I'm using now.

mmcc commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
-  this.el_ = null;
-};
+    // Remove element from DOM
+    if (this.el_.parentNode) {
```
Should we use the `#el()` method instead of the "private" var when just reading this (throughout the module)?

heff replied:
The official Closure Library style is to use the private references everywhere internally so compiler can mangle them, and then export the methods for anything you want accessible externally. That's not always obvious for other contributors, and I also started to worry that people would start using the private properties externally after seeing internal code, but then on the other hand I think uglify has an option to mangle private props now if we wanted to keep that part of the file size savings.

Either way it's open for a change.

mmcc commented on src/js/control-bar/duration-display.js:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  createEl() {
+    let el = Component.prototype.createEl.call(this, 'div', {
```
`super.createEl`?

mmcc commented on src/js/control-bar/fullscreen-toggle.js:
```diff
@@ -8,33 +7,25 @@ import Button from '../button';
 
-Component.registerComponent('FullscreenToggle', FullscreenToggle);
+  buildCSSClass() {
+    return 'vjs-fullscreen-control ' + Button.prototype.buildCSSClass.call(this);
```
`super`

mmcc commented on src/js/control-bar/mute-toggle.js:
```diff
@@ -29,50 +29,51 @@ var MuteToggle = Button.extend({
-  });
-};
+  createEl() {
+    return Button.prototype.createEl.call(this, 'div', {
```
![super](http://vignette1.wikia.nocookie.net/cartoonfatness/images/0/03/Superman_john_byrne1.png/revision/latest?cb=20140114011034)

mmcc commented on src/js/control-bar/mute-toggle.js:
```diff
@@ -29,50 +29,51 @@ var MuteToggle = Button.extend({
-  } else if (vol < 0.67) {
-    level = 2;
-  }
+    if (vol === 0 || this.player_.muted()) {
```
This might be outside of the scope of this PR, but perhaps this would be more readable as a switch statement.

gkatsev replied:
This doesn't translate to a switch since they're inequalities and not concrete values.
Also, switch is just a glorified if/else.
What about?
```js
level = (vol === 0)  ? 0 :
        (vol < 0.33) ? 1 :
        (vol < 0.67) ? 2 ;
```
or maybe
```js
if (vol === 0 || this.player_.muted()) {
  level 0;
} else
if (vol < 0.33) {
  level 1;
} else
if (vol < 0.67) {
  level = 2;
}
```
Or maybe
```js
if (vol === 0 || this.player_.muted()) {
  level = 0;                           } else
if (vol < 0.33) {
  level = 1;                           } else
if (vol < 0.67) {
  level = 2;                           }
```

gkatsev replied:
I'm actually really liking this look for this particular usecase:
```js
level = (vol === 0)  ? 0 :
        (vol < 0.33) ? 1 :
        (vol < 0.67) ? 2 ;
```

heff replied:
That last one does look really clean. I think it depends if we want to be more strict in our linting about ternaries to avoid things like in the other PR.

(I'm not going to try to fix this one in this PR)

gkatsev replied:
Ternaries are very useful sometimes, like here. They shouldn't definitely not be overused, but they shouldn't be outright discounted.
If we feel like the style here is more readable, we should go with it even if our current linting configuration may complain.

mmcc commented on src/js/control-bar/play-toggle.js:
```diff
@@ -9,45 +8,45 @@ import * as Lib from '../lib';
-  return 'vjs-play-control ' + Button.prototype.buildCSSClass.call(this);
-};
+  buildCSSClass() {
+    return 'vjs-play-control ' + Button.prototype.buildCSSClass.call(this);
```
![super](http://www.theukulelereview.com/wp-content/uploads/2011/02/super-mario-bros1.jpg?w=172)

mmcc commented on src/js/control-bar/playback-rate-menu/playback-rate-menu-button.js:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  createEl() {
+    let el = MenuButton.prototype.createEl.call(this);
```
![super](http://img2.wikia.nocookie.net/__cb20110405210554/muppet/images/7/73/Super-Grover-2.0.png)

mmcc commented on src/js/control-bar/playback-rate-menu/playback-rate-menu-button.js:
```diff
@@ -0,0 +1,108 @@
+
+  onClick() {
+    // select next rate option
+    let currentRate = this.player().playbackRate();
```
We definitely need to be more consistent. We use `this.player_` and `this.player()` seemingly at random.

mmcc commented on src/js/control-bar/progress-control/seek-bar.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  createEl() {
+    return Slider.prototype.createEl.call(this, 'div', {
```
![super - I'm starting to get pretty weak on these images](http://details-of-cars.com/wp-content/uploads/parser/Dodge-super-bee-6.gif)

mmcc commented on src/js/control-bar/progress-control/seek-bar.js:
```diff
@@ -0,0 +1,93 @@
+
+}
+
+SeekBar.prototype.options_ = {
```
Any reason we're adding this to the prototype instead of defining it in the constructor?

gkatsev replied:
Static properties like this can be shared across instances as opposed to being created each time in the constructor. Whether we want to bring it into the constructor will depend on what we want out of these options.

heff replied:
Yeah, setting them on the prototype makes them globally overridable, which I assume is handy at the player level for setting global options across players, but less useful in other components.

I've started to think the same thing about moving the options to the constructor, and also potentially killing the `options()` accessor. It kind of creates two sources of truth. Not sure how much people are relying on that now though.

mmcc commented on src/js/control-bar/text-track-controls/captions-button.js:
```diff
@@ -0,0 +1,49 @@
+
+  update() {
+    let threshold = 2;
+    TextTrackButton.prototype.update.call(this);
```
![super...bowl?](http://upload.wikimedia.org/wikipedia/en/thumb/d/d7/Super_Bowl_XXVI_Logo.svg/1280px-Super_Bowl_XXVI_Logo.svg.png)

mmcc commented on src/js/control-bar/volume-control/volume-bar.js:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  createEl() {
+    return Slider.prototype.createEl.call(this, 'div', {
```
![super](http://upload.wikimedia.org/wikipedia/en/6/66/Super_Poster.jpg)

mmcc commented on src/js/control-bar/volume-control/volume-control.js:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  createEl() {
+    return Component.prototype.createEl.call(this, 'div', {
```
![super](http://www.wpclipart.com/education/encouraging_words/super_Star.png)

mmcc commented on src/js/poster-image.js:
```diff
@@ -0,0 +1,101 @@
+   */
+  dispose() {
+    this.player().off('posterchange', this.update);
+    Button.prototype.dispose.call(this);
```
![super](http://i.ytimg.com/vi/rZOeizpBsn4/maxresdefault.jpg)

mmcc commented on src/js/tech/html5.js:
```diff
@@ -0,0 +1,682 @@
+
+  dispose() {
+    Html5.disposeMediaElement(this.el_);
+    Tech.prototype.dispose.call(this);
```
`super`

heff replied:
I'm sad you gave up on the super images. :( In my email they got progressively larger as I scrolled down. :-P

mmcc commented on src/js/tech/html5.js:
```diff
@@ -0,0 +1,682 @@
+
+  removeRemoteTextTrack(track) {
+    if (!this['featuresNativeTextTracks']) {
+      return Tech.prototype.removeRemoteTextTrack.call(this, track);
```
`super`

mmcc commented on src/js/tech/html5.js:
```diff
@@ -0,0 +1,682 @@
+
+  addRemoteTextTrack(options) {
+    if (!this['featuresNativeTextTracks']) {
+      return Tech.prototype.addRemoteTextTrack.call(this, options);
```
`super`

mmcc commented on src/js/tech/html5.js:
```diff
@@ -0,0 +1,682 @@
+  }
+  addTextTrack(kind, label, language) {
+    if (!this['featuresNativeTextTracks']) {
+      return Tech.prototype.addTextTrack.call(this, kind, label, language);
```
`super`

mmcc commented on src/js/tech/html5.js:
```diff
@@ -0,0 +1,682 @@
+
+  textTracks() {
+    if (!this['featuresNativeTextTracks']) {
+      return Tech.prototype.textTracks.call(this);
```
`suuuper`

mmcc commented on src/js/tech/tech.js:
```diff
@@ -0,0 +1,536 @@
+
+    if (this.manualTimeUpdates) { this.manualTimeUpdatesOff(); }
+
+    Component.prototype.dispose.call(this);
```
`superduper`

mmcc commented on src/js/tracks/text-track-display.js:
```diff
@@ -0,0 +1,185 @@
+  }
+
+  createEl() {
+    return Component.prototype.createEl.call(this, 'div', {
```
`super`

mmcc commented:
I'll take another pass tomorrow, but first pass showed a few things that can refactored to `super` calls.

Just a note not necessarily related to this PR: things seem a bit inconsistent about when we use things like `el_` and `player_` versus their respective getter methods.

gkatsev commented on src/js/component.js:
```diff
@@ -3,7 +3,7 @@
  */
 
-import CoreObject from './core-object.js';
+// import CoreObject from './core-object.js';
```
remove if not needed.

gkatsev commented on src/js/component.js:
```diff
@@ -38,14 +38,16 @@ import window from 'global/window';
+
+    // The component might be the player itself and we can't pass `this` to super
+    if (!player && this.play) {
+      this.player_ = player = this;
```
Does the player not get the player argument?
I wonder if we can use default args here:
```js
class Component {
  constructor (player = this, options, ready) {
    this.player_ = player;
  }
}
```

heff replied:
Yeah, I can't pass `this` (the player) as the first arg to the component constructor or I get yelled at for using `this` before the super().

We could default to this, but I also think it could be good to throw an error if there's no player arg when other components are created. If a component depends on the player API, it will throw random errors later.

gkatsev commented on src/js/component.js:
```diff
@@ -65,14 +67,20 @@ var Component = CoreObject.extend({
 
     // Create element if one wasn't provided in options
-    this.el_ = options['el'] || this.createEl();
+    if (options['el']) {
```
`.el`

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
-Component.prototype.dispose = function(){
-  this.trigger({ type: 'dispose', 'bubbles': false });
+  // Temp for ES6 class transition, remove before 5.0
+  init() {
```
when this PR is merged in, we probably want to start on the videojs-migrate plugin.

heff replied:
Yeah, definitely.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+
+    // Dispose all children.
+    if (this.children_) {
+      for (var i = this.children_.length - 1; i >= 0; i--) {
```
if we are making es5-shim a requirement, I would love to turn all for-loops into iterators.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   *
+   * @return {Element}
+   */
+  contentEl(){
```
what's the difference between `contentEl` and `el`? Should be have both? Why not just insert children into `el`?

heff replied:
By default they point to the same, but some elements use sub elements to create their appearance and you don't want the the children to go into the main element anymore. MenuButtons might be one of those but I'm not remembering exactly.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   * @suppress {accessControls|checkRegExp|checkTypes|checkVars|const|constantProperty|deprecated|duplicate|es5Strict|fileoverviewTags|globalThis|invalidCasts|missingProperties|nonStandardJsDocs|strictModuleDepCheck|undefinedNames|undefinedVars|unknownDefines|uselessCode|visibility}
+   */
+  addChild(child, options){
+    let component, componentName;
```
`componentName` shouldn't be defined out here.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
-    this.childNameIndex_[componentName] = component;
-  }
+    var childFound = false;
+    for (var i = this.children_.length - 1; i >= 0; i--) {
```
should be a `let`

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   * @param  {Function}             third   The event handler
+   * @return {vjs.Component}        self
+   */
+  on(first, second, third){
```
it would be really nice to name these better names

heff replied:
You see why it's difficult though, yeah?

gkatsev replied:
yes. What about: `first -> typeOrComponent`, `second -> `handlerOrType`, and `third -> handler`.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   * @param  {Function} fn Ready listener
+   * @return {vjs.Component}
+   */
+  ready(fn){
```
I had the thing below, but now I realize it isn't quite what we want, since we do only want to add the function to the queue only if it was given to us.
```js
class Component {
  ready (fn = function() {}) {
    if (this.isReady_) {
      fn.call(this);
    } else {
      this.readyQueue_ = (this.readyQueue_ || []);
      this.readyQueue_.push(fn);
    }
}
```

heff replied:
Cool. I added the other cleanup item.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+   * @private
+   */
+  dimension(widthOrHeight, num, skipListeners){
+    if (num !== undefined) {
```
`num` can be set to a default argument of `0`, then we only need to check if it's `null` or `NaN`.

heff replied:
But then we wouldn't know if dimension is being used as a getter.

gkatsev commented on src/js/component.js:
```diff
@@ -82,1182 +90,1103 @@ var Component = CoreObject.extend({
+      } else if (num === 'auto') {
+        this.el_.style[widthOrHeight] = '';
+      } else {
+        this.el_.style[widthOrHeight] = num+'px';
```
this should *not* be turned into a template string.

mmcc replied:
[it's not](https://github.com/mmcc/video.js/blob/es6-defaults-interpolation/src/js/component.js#L981)

gkatsev replied:
Yeah, not saying that it's an issue. Just that we should be careful around things like these because `num` is user inputtable and if it's carefully crafted, could cause issues with template strings since they have access to the variables.

mmcc replied:
Ah good point. Noted for future reference.

gkatsev commented on src/js/control-bar/current-time-display.js:
```diff
@@ -0,0 +1,42 @@
+  updateContent() {
+    // Allows for smooth scrubbing, when player can't keep up.
+    let time = (this.player_.scrubbing) ? this.player_.getCache().currentTime : this.player_.currentTime();
+    this.contentEl_.innerHTML = '<span class="vjs-control-text">' + this.localize('Current Time') + '</span> ' + Lib.formatTime(time, this.player_.duration());
```
this will probably be fun to rebase with the template strings PR

gkatsev commented on src/js/control-bar/mute-toggle.js:
```diff
@@ -29,50 +29,51 @@ var MuteToggle = Button.extend({
+    // Don't rewrite the button text if the actual text doesn't change.
+    // This causes unnecessary and confusing information for screen reader users.
+    // This check is needed because this function gets called every time the volume level is changed.
+    if(this.player_.muted()){
```
```js
let toMute = this.player_.muted() ? 'Unmute' : 'Mute';
let localizedMute = this.localize(toMute);
if (this.el_.children[0].children[0].innerHTML !== localizedMute) {
  this.el_.children[0].children[0].innerHTML = localizedMute;
}
```

heff replied:
Much nicer. Updated.

gkatsev commented on src/js/tracks/text-track.js:
```diff
@@ -227,7 +226,7 @@ TextTrack.prototype.removeCue = function(removeCue) {
  */
-let parseCues = function(srcContent, track) {
+var parseCues = function(srcContent, track) {
   if (typeof window['WebVTT'] !== 'function') {
```
why was this changed? We should move `parseCues` up in the file, if the TDZ is the issue.

heff replied:
Yeah, can't use a let var before it's defined in the file. How about I make it `function parseCues()`? I don't like burying the constructor in its own file.

heff commented:
I guess the punishment for having a crappy diff and forcing people to look at every line of code is you get to address every style problem from ever :) There's at least fewer than I'd expect.

heff commented:
Updated with the super calls and cleanup changes mentioned. I've moved some not directly related discussions to their own issues. Otherwise the important bits here are done so going to merge this in so we can move forward.
